src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,197 @@
+                    return ParseResult.Incomplete;
+                }
+
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                if (line.Length <= 1)
+                {
+                    throw new FormatException("There was an error in the frame format");
+                }
+
+                if (IsMessageEnd(line))
+                {
+                    _internalParserState = InternalParseState.ReadEndOfMessage;
+                }
+                // To ensure that the \n was preceded by a \r
+                // since messages can't contain \n.
+                // data: foo\n\bar should be encoded as
+                // data: foo\r\n
+                // data: bar\r\n
+                else if (line[line.Length - _sseLineEnding.Length] != ByteCR)
+                {
+                    throw new FormatException("A '\\n' character can only be used as a line ending");
+                }
+
+                switch (_internalParserState)
+                {
+                    case InternalParseState.ReadMessageType:
+                        messageType = GetMessageType(line);
+
+                        _internalParserState = InternalParseState.ReadMessagePayload;
+
+                        start = lineEnd;
+                        consumed = lineEnd;
+                        break;
+                    case InternalParseState.ReadMessagePayload:
+                        EnsureStartsWithDataPrefix(line);
+
+                        // Slice away the 'data: '
+                        var payloadLength = line.Length - (_dataPrefix.Length + _sseLineEnding.Length);
+                        var newData = line.Slice(_dataPrefix.Length, payloadLength).ToArray();
+                        _data.Add(newData);
+
+                        start = lineEnd;
+                        consumed = lineEnd;
+                        break;
+                    case InternalParseState.ReadEndOfMessage:
+                        if (_data.Count > 0)
+                        {
+                            // Find the final size of the payload
+                            var payloadSize = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                payloadSize += dataLine.Length + _newLine.Length;
+                            }
+
+                            // Allocate space in the paylod buffer for the data and the new lines 
+                            var payload = new byte[payloadSize - _newLine.Length];
+
+                            var offset = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                dataLine.CopyTo(payload, offset);
+                                offset += dataLine.Length;
+                                if (_data.Count > 1 && offset < payload.Length)
```

Comment:
I don't think you need `_data.Count > 1` here. If there is only one item, `offset` should equal `payload.Length` now.